### DISCUSSION TITLE
[이유경 멘토] 3주차 멘토링 정리 😎

### DISCUSSION THREAD

**Discussion by yukyunglee** [bcaitech1/p4-dkt-freshtomato#91] [General]
### 6.8 멘토링 정리

1. 진행사항 공유

   * 아직 정확한 메인 모델은 없음
   * ML (재훈, 유라) : LBGM
     * **3가지 Boost 앙상블 추천 (츄라이 츄라이 ~ 👀) : https://www.kaggle.com/samratp/lightgbm-xgboost-catboost **
   * DL  (정현, 진현, 아라, 민용) : LSTM, SAINT, LASTQuery, GRU
     * Saint는 생각보다 높은 성능이 나오진 않았음 (민용님 피셜)
       * **추가적인 테스트가 필요해보입니다 ! GRU와 똑같은 feature를 사용해서 비교해보세요 😎**
       * Encoder, Decoder의 feature가 달라지거나 같아지는건  결국 엄청나게 성능을 향상시키는 문제는 아닙니다. (성능이 좋다고 하더라도 미미한 개선이 이루어질거에요)

2. 이슈 1 : Concat에 대하여..

   * feature를 결합하는 방식에 대한 논의

     * 역시나 Multimodal transformer의 방법 (feature들을 selfattention으로 계산)을 사용하면 성능이 좋아집니다..만 기회비용이 너무너무 큰 작업이에요..😂

   * 하 ! 지 ! 만 ! 또 다른 방법을 생각해본다면 Convolution을 태우는 방법도 있겠네요 !?(갑자기 생각났습니다ㅋㅋ)

     * 그래서 찾아보니 비슷한 논문이 있네요 ? (첨부파일로 드릴게요)

       * **그래서 추천하느냐,, ? : 잘 모르겠습니다 ..🤨 띵크빅이라는 생각은 안드는 방법이네용..** 

       
![image](https://user-images.githubusercontent.com/35751880/121141490-f6aa0b00-c875-11eb-8450-4cf8b095340b.png)

   * 역시나 Linear layer ~ Concat이 가장 일반적인 방법으로 보입니다
   * 그럼에도 불구하고 더 많은 정보를 찾고싶으시다면 'Feature fusion' 으로 검색해서 찾아보세요 !

3. 이슈 2 : 앙상블에 대하여

   * 슬슬 앙상블 스코어를 뽑아보시고 어떤 모델을 버릴지 선택을 시작해보세요 !
   * 하드보팅 vs 소프트 보팅
     * 이건 다 해보고 결정하는게 좋습니다
     * 다만, 동일한 스코어를 보일 때 어떤 기준으로 정답을 선택할지 결정하는게 좋습니다 
       * 예를들어 3:3인 경우엔 , 가장 성능이 좋은 모델의 솔루션을 택한다 
     * 모델이 '덜' 학습되거나 좋지 않은 경우엔 소프트 보팅이 더 안좋은 결과를 초래 할 수 있습니다

4. 이슈 3: stratified 기준

   * 랜덤으로 진행 
   * 일정 기준으로 진행중 (민용님)
     * 마지막 문제를 맞췄는가에 따라서 나누었음
       * 매우 좋은 기준이라 생각합니다. 추가적으로 문제의 난이도에 대해 같이 고려해서 나눠보시는것도 추천드려요

5. 이슈 4 : Augmentation

   * 250씩 잘라서 데이터로 사용
     * 성능이 안좋음
       * 문제 난이도의 분포를 한번 확인해보세요 ! 
       * 난이도 차이가 너무 많이 난다면 (분포가 많이 다르다면) 좋은 augmentation이 아닐 가능성이 높습니다. 
       * **기존 데이터와 비슷한 분포를 가지도록 나누어보시면 어떨까요 ? (뇌피셜입니다..)**

6. 이슈 5 : 추가적인 feature에 대한 Trick

   * Continuous Feature들을 categorical로 바꾸어서 성능이 향상되었다면 동일하게 categorical의 Class를 의미있는 text로 바꾸시고 PLM(Pretrained language model)에서 Embedding 값을 가져와서 사용해보세요 !
     * class의 meaning과 Context를 모델이 알 수 있도록 강제할 수 있는 방법 중 하나입니다
     * 사실은 점수 향상을 위한 엄청난 꼼수...🤔 
     * 요것도 막상 해보면 안될수도 있지만.. 쉽게 적용할수 있는 방법이니 추천 !
   * 중요 Feature를 시퀀스 마다 concat해보세요 ! 
     * 시퀀스를 잘 맞춰서 해주어야합니다
       * t번째 feature + 중요 feature 여러번 붙여주기 
     * encoder - decoder 구조에서 잘 working했던 경험이 있습니다

7. 짧은 진로상담
 * 슬랙 참고

8. 책 추천 (상단의 두가지책은 무조건...)
   * 파이썬 알고리즘 인터뷰 http://www.yes24.com/Product/Goods/91084402
   * 파이썬 자료구조 알고리즘 http://www.yes24.com/Product/Goods/74971408
   * 김기현의 자연어 처리 딥러닝 캠프 파이토치 http://m.yes24.com/Goods/Detail/74802622

9. 추가 멘토링

   1) 11일 4시반 (마지막 멘토링)

   2) 원하신다면 17일 멘토링도 !

   3) 4번 멘토링 진행은 운영진 분들께 말씀드렸으니 문제 없을듯 합니다 ~~

**Comment by yukyunglee** [comment]
[Deep_Learning_Based_Weighted_Feature_Fusion_Approa.pdf](https://github.com/bcaitech1/p4-dkt-freshtomato/files/6614283/Deep_Learning_Based_Weighted_Feature_Fusion_Approa.pdf)

**Comment by gvsteve24** [comment]
정리 감사드립니다:)

**Reply by yukyunglee** [reply]
@여러분 saint 집중해서 짜보시는건 어떨까요 ? DL팀에서 2분정도 집중해서 빠르게 실험해보는게 좋을 것 같다는 생각이..!

**Reply by arabae** [reply]
우선은 저희가 다양한 모델을 계속 시도하고, 마지막에 앙상블을 목표로 하고있어서 개개인이 모두 다른 모델로 실험을 유지할 것 같습니다! 진현님이 SAINT 모델을 집중해서 보실거에요 :)
➕ GRUATTN에서 좋았던 feature, hyperparameter 등을 가져와서 사용해봐도 조금 성능이 떨어졌지만 계속해서 실험은 같이 진행할 것입니다!
그리고, transformer 계열 모델 (SAINT, LastQuery)은 Riiid 데이터로 pretrain을 진행한 뒤, 대회 데이터로 retrain을 진행할 예정입니다!

올려주신 Conv layer를 활용해서 categorical과 continuous한 feature를 합치는 방법도 꼭 실험해보겠습니다!!

**Reply by yukyunglee** [reply]
네 ! 가능한 capa 내에서 다양하게 시도해보시고, Saint 코드를 빠르게 보니 민용님이 말씀하신 것 처럼 원래 코드랑 좀 다르더라구요! 이걸 원래대로 고치는것도 하나의 개선 포인트가 아닐까 싶습니다. Riiid 데이터로 pretrain을 한다는 것은 미리 학습한 다음에 대회 데이터로 추가 학습을 하신다는 의미 맞을까요 ? (오래 걸리지 않는다면 가장 좋은 모델에 Riiid 데이터로 학습부터 해보시는걸 추천드립니다. 그 이유는 두 데이터의 분포가 매우 상이할 경우 학습 성능이 크게 저하할 수 있습니다)

**Reply by gvsteve24** [reply]
네, pretrain은 말씀해주신 의미가 맞습니다.!

**Reply by arabae** [reply]
- SAINT 코드를 논문과 같이 바꿔서 실험도 진행해봤습니다! (interaction만 decoder에 넣어주는 방식) 이때, 성능이 개선되지는 않았지만 비교실험으로 계속 가져가도 좋을 것 같습니다!
- LastQuery model에 pretrain을 진행하였을 때 5시간 소요되었습니다. GRU가 현재까지 가장 좋았지만, 작은 모델이기 때문에 저희도 성능이 저하될 수 있을거라 생각이 들어서 우선 Transformer를 사용하는 모델에 적용해보기로 하였습니다!

**Reply by yukyunglee** [reply]
네 좋습니다 ㅎㅎ 데이터가 많아서 그런지 꽤 오랜시간이 걸리는군요 ! 아라님이 잘 이끌어주고 계시다는 생각이 듭니다. Riiid Training 후 추가 학습을 진행할 경우 성능이 도출되면 알려주세요 ㅎㅎ !
